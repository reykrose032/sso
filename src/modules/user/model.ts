import { getModelForClass } from '@typegoose/typegoose';
import { ObjectId } from 'mongodb';
import * as bcrypt from 'bcrypt';

import { User } from '../../entities/user';
import { NewUserInput } from './input';
import { UserResponse } from './response';

export const UserMongooseModel = getModelForClass(User);

export default class UserModel {
    async getById(_id: ObjectId): Promise<User | null> {
        return UserMongooseModel.findById(_id).lean().exec();
    }

    async getAll(): Promise<User[] | null> {
        return UserMongooseModel.find().lean().exec();
    }

    async authenticate(authInput: NewUserInput): Promise<UserResponse> {
        const { name, password } = authInput;
        const userFromDB = await UserMongooseModel.findOne({ name }).lean().exec();

        if (!userFromDB)
            throw new Error('User not found');
        
        const { hashPassword, _id, token } = userFromDB;

        if (!await bcrypt.compare(password, hashPassword))
            throw new Error('Wrong password');

        return { _id, name, token };
    }

    async create(input: NewUserInput): Promise<UserResponse> {
        const { name, password } = input;
        const userFromDB = await UserMongooseModel.findOne({ name }).lean().exec();
        if (userFromDB)
            throw new Error('Username taken');

        const saltPassword = await bcrypt.genSalt();
        const hashPassword = await bcrypt.hash(password, saltPassword);
        const token = await bcrypt.hash(String(Date.now()), saltPassword);
        const user = new UserMongooseModel({ name, saltPassword, hashPassword, token });

        const { _id } = await user.save();

        return { _id, name, token };
    }
}