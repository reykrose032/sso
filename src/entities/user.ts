import { ObjectType, Field } from "type-graphql";
import { prop } from "@typegoose/typegoose";
import { ObjectId } from "mongodb";

@ObjectType()
export class User {
  @Field()
  readonly _id!: ObjectId;

  @prop()
  @Field()
  name!: string;

  @prop()
  @Field()
  hashPassword!: string;
  
  @prop()
  @Field()
  saltPassword!: string;

  @prop()
  @Field()
  token!: string;
}
