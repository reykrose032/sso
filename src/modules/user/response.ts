import { Field, ObjectType } from "type-graphql";
import { MaxLength, MinLength } from "class-validator";
import { ObjectId } from 'mongodb';
import { prop } from "@typegoose/typegoose";

@ObjectType()
export class UserResponse {
    @Field()
    readonly _id!: ObjectId;

    @Field()
    @prop()
    name: string;
}
