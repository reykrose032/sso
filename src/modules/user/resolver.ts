import { Resolver, Arg, Query, Mutation, ID } from "type-graphql";
import { Service } from "typedi";
import { ObjectId } from "mongodb";

import { User } from "../../entities";
import UserService from "./service";
import { NewUserInput } from "./input";
import { UserResponse } from "./response";

/*
  IMPORTANT: Your business logic must be in the service!
*/

@Service() // Dependencies injection
@Resolver((of) => User)
export default class UserResolver {
  constructor(private readonly userService: UserService) {}

  @Query((returns) => User)
  async getUser(@Arg("id") id: ObjectId) {
    const user = await this.userService.getById(id);

    return user;
  }

  @Query(() => [User])
  async getAllUsers() {
      const users = await this.userService.getAllUsers();
      return users;
  }

  @Mutation(() => UserResponse)
  async authenticate(
    @Arg('authenticateData') authenticateData: NewUserInput
  ): Promise<UserResponse> {
    const user = await this.userService.authenticate(authenticateData);
    return user;
  }

  @Mutation((returns) => UserResponse)
  async createUser(
    @Arg("createUserData") createUserData: NewUserInput
  ): Promise<UserResponse> {
    const user = await this.userService.addUser(createUserData);
    return user;
  }
}
