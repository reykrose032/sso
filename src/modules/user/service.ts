import { Service } from "typedi";
import { ObjectId } from "mongodb";

import UserModel from "./model";
import { User } from "../../entities";
import { NewUserInput } from "./input";
import { UserResponse } from "./response";

@Service() // Dependencies injection
export default class TodoService {
  constructor(private readonly userModel: UserModel) {}

  public async getById(_id: ObjectId): Promise<User | null> {
    return this.userModel.getById(_id);
  }

  public async getAllUsers(): Promise<User[] | null> {
      return this.userModel.getAll();
  }

  public async addUser(input: NewUserInput): Promise<UserResponse> {
    const newUser = await this.userModel.create(input);

    // Business logic goes here
    // Example:
    // Trigger push notification, analytics, ...

    return newUser;
  }

  public async authenticate(authInput: NewUserInput): Promise<UserResponse> {
    return this.userModel.authenticate(authInput);
  }
}
