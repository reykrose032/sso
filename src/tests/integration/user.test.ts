import { buildSchema } from '../../utils';
import { UserMongooseModel } from '../../modules/user/model'

import {
    connect,
    clearDatabase,
    closeDatabase,
    populateDatabase,
} from "../utils";
import { ApolloServer, gql } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';

beforeAll(async () => connect());

// beforeEach(async () => {
//     await populateDatabase(UserMongooseModel, [
//         {
//             name: 'Manolo'
//         },
//         {
//             name: 'XDD'
//         }
//     ]);
// });

// afterEach(async () => {
//     await clearDatabase();
// });

afterAll(async (done) => {
    await closeDatabase();
    done();
});

describe('User', () => {
    it(`should create an user`, async () => {
        const graphQLSchema = await buildSchema();

        const server = new ApolloServer({
            schema: graphQLSchema,
        }) as any;

        const { mutate } = createTestClient(server);

        const variables = {
            createUserData: {
                name: `Test name`,
                password: `123`
            },
        };

        const CREATE_USER = gql`
        mutation createUser($createUserData: NewUserInput!) {
            createUser(createUserData: $createUserData) {
                name
            }
        }
        `;

        const res = await mutate({
            mutation: CREATE_USER,
            variables,
        });

        // expect(res.data).toMatchObject({
        //     "createUser": {
        //     "name": "Test name",
        //   }
        // });

        // expect(res.data).toMatchSnapshot({
        //     createUser: {
        //         name: 'Test name'
        //     }
        // });

        expect(res).toMatchSnapshot();
    });

    // it(`should get all users`, async () => {
    //     const graphQLSchema = await buildSchema();

    //     const server = new ApolloServer({
    //         schema: graphQLSchema,
    //     }) as any;

    //     const { query } = createTestClient(server);

    //     const GET_USERS = gql`
    //     query {
    //         getAllUsers {
    //           _id
    //           name
    //         }
    //       }
    //     `;

    //     const res = await query({
    //         query: GET_USERS
    //     });

    //     // expect(res.data).toMatchObject({
    //     //     "users": {
    //     //     "name": "Test name",
    //     //   }
    //     // });

    //     expect(res).toMatchSnapshot();
    // })

    it(`should authenticate`, async () => {
        const graphQLSchema = await buildSchema();

        const server = new ApolloServer({
            schema: graphQLSchema,
        }) as any;

        const { mutate } = createTestClient(server);

        const variables = {
            authenticateData: {
                name: `Test name`,
                password: `123`
            },
        };

        const AUTHENTICATE = gql`
        mutation authenticate($authenticateData: NewUserInput!) {
            authenticate(authenticateData: $authenticateData) {
                _id
                name
                token
            }
        }
        `;

        const res = await mutate({
            mutation: AUTHENTICATE,
            variables,
        });

        expect(res).toMatchSnapshot();
    })

    it(`should not authenticate`, async () => {
        const graphQLSchema = await buildSchema();

        const server = new ApolloServer({
            schema: graphQLSchema,
        }) as any;

        const { mutate } = createTestClient(server);

        const variables = {
            authenticateData: {
                name: `Test name`,
                password: `12345`
            },
        };

        const AUTHENTICATE = gql`
        mutation authenticate($authenticateData: NewUserInput!) {
            authenticate(authenticateData: $authenticateData) {
                _id
                name
                token
            }
        }
        `;

        const res = await mutate({
            mutation: AUTHENTICATE,
            variables,
        });

        expect(res).toMatchSnapshot();
    })
});