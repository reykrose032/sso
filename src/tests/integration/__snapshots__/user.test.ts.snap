// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User should authenticate 1`] = `
Object {
  "data": Object {
    "authenticate": Object {
      "_id": "60be9f1da5a6f413780d8ea4",
      "name": "Test name",
      "token": "$2b$10$elVSdQ0yDyo4x8mv.7iLneP5Bt4amZ8.FSefFHTdf9RjrlYWViZE.",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`User should create an user 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "name": "Test name",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`User should not authenticate 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Wrong password],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
